pkg/scheduler/partition.go

```diff
@@ -338,6 +338,32 @@ func (pc *PartitionContext) AddApplication(app *objects.Application) error {
 		return fmt.Errorf("failed to find queue %s for application %s", queueName, appID)
 	}
 
+	// Has user or group defined with any limits? If yes, Is there a room to operate?
+	user := queue.GetUserGroupManager().GetUser(app.GetUser().User)
+	if user != nil {
+		if user.CanRun() {
+			user.IncRunningApplications()
+		} else {
+			return fmt.Errorf("user '%s' has reached max applications limit in queue %s", app.GetUser().User,
+				queueName)
+		}
+	}
+
+	for _, group := range app.GetUser().Groups {
+		// Is there any group (to which user belongs to) config has limit settings?
+		g := queue.GetUserGroupManager().GetGroup(group)
+		if g != nil && g.CanRun() {
+			g.IncRunningApplications()
+			if user != nil {
+				user.SetUsedGroup(g.GetName())
```

Comment:
What if the user belongs to multiple groups? Can that happen?
If so, is the list of groups predictable?

Reply:
Yes, a user may belong to more than group. Added user.SetUsedGroup() mainly to reduce this kind of friction by storing the group and use the same in future as well. As and when configured group arrives while traversing groups of user, that group would be picked up and set. Also, setting group using user.SetUsedGroup() avoids parsing groups again while removing the application to larger extent.